ArdusatSDK.cpp

```diff
@@ -453,266 +217,1101 @@ const char * valuesToCSV(const char *sensorName, unsigned long timestamp, int nu
+Magnetic::Magnetic(void) :
+  gaussScale(LSM303_MAG_SCALE4GAUSS)
 {
-  _resetOutBuf();
-  _writeJSONValue(_getOutBuf(), sensor_name, unit_to_str(temp.header.unit), temp.t);
-  return _getOutBuf();
+  this->initializeHeader(SENSORID_ADAFRUIT9DOFIMU, DATA_UNIT_MICROTESLA, magnetic_sensor_name);
 }
 
-const char * luminosityToJSON(const char *sensor_name, luminosity_t & lum)
+/**
+ * @brief   Sets the gauss scale configuration variable
+ * @ingroup magnetic
+ *
+ * @param gaussScale Advanced configuration for magnetometer's scale
+ *     - LSM303_MAG_SCALE1_3GAUSS
+ *     - LSM303_MAG_SCALE2GAUSS
+ *     - LSM303_MAG_SCALE2_5GAUSS
+ *     - LSM303_MAG_SCALE4GAUSS (Default)
+ *     - LSM303_MAG_SCALE4_7GAUSS
+ *     - LSM303_MAG_SCALE5_6GAUSS
+ *     - LSM303_MAG_SCALE8GAUSS
+ *     - LSM303_MAG_SCALE12GAUSS
+ */
+void Magnetic::configure(lsm303_mag_scale_e gaussScale) {
+  this->gaussScale = gaussScale;
+  this->initialized = false;
+}
+
+/**
+ * @brief   Initializes the sensor with any set configurations or defaults
+ * @ingroup magnetic
+ *
+ * @retval true  Successfully initialized
+ * @retval false Failed to initialize
+ */
+boolean Magnetic::initialize(void) {
+  return lsm303_mag_init(this->gaussScale);
+}
+
+/**
+ * @brief   Takes a reading from the sensor
+ * @ingroup magnetic
+ */
+void Magnetic::read(void) {
+  this->header.timestamp = millis();
+  lsm303_getMag(&(this->x), &(this->y), &(this->z));
+}
+
+/**
+ * @brief   Returns last read value in CSV format
+ * @ingroup magnetic
+ * @param   sensorName The text to display next to the value
+ * @return  sensor readings in CSV format or empty string if uninitialized
+ */
+const char * Magnetic::toCSV(const char * sensorName) {
+  if (this->initialized) {
+    return valuesToCSV(sensorName, this->header.timestamp, 3,
+                       this->x, this->y, this->z);
+  } else {
+    return "";
+  }
+}
+
+/**
+ * @brief   Returns last read value in JSON format
+ * @ingroup magnetic
+ * @param   sensorName The text to display next to the value
+ * @return  sensor readings in JSON format or empty string if uninitialized
+ */
+const char * Magnetic::toJSON(const char * sensorName) {
+  if (this->initialized) {
+    return valuesToJSON(sensorName, this->header.unit, 3, "X", this->x,
+                              "Y", this->y, "Z", this->z);
+  } else {
+    return "";
+  }
+}
+
+
+/**************************************************************************//**
+ * @brief   Constructs Orientation calculation object
+ * @ingroup orientation
+ *
+ * Example Usage:
+ * @code
+ *     Orientation orient;                               // Instantiate sensor object
+ *     orient.begin();                                   // Initialize sensor
+ *     Serial.println(orient.readToJSON("orientation")); // Read and print values in JSON
+ * @endcode
+ *****************************************************************************/
+Orientation::Orientation(void) {
+  Acceleration accel;
+  Magnetic mag;
```

Comment:
I should add that the "right"/canonical C++ way to do this would be to use `new` (heap-allocate), but store them in smart pointers (http://www.umich.edu/~eecs381/handouts/C++11_smart_ptrs.pdf), which grew out of Boost's old reference counted pointer library and were made part of the standard library in C++11. 

But the Arduino IDE doesn't doesn't have C++11 support pre 1.6.x without compiler flags (and I don't think 1.0.x has it available at all), so we can't really do that here...